analyzers/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs

```diff
@@ -52,10 +52,16 @@ public abstract class PureAttributeOnVoidMethodBase<TSyntaxKind> : SonarDiagnost
                 SymbolKind.Method);
 
         protected static AttributeData InvalidPureDataAttributeUsage(ISymbol symbol) =>
-            symbol is IMethodSymbol { ReturnsVoid: true } methodSymbol
-                   && !methodSymbol.Parameters.Any(p => p.RefKind != RefKind.None)
+            symbol is IMethodSymbol methodSymbol
+                   && (ReturnsVoid(methodSymbol) || ReturnsTask(methodSymbol))
                    && methodSymbol.GetAttributes().FirstOrDefault(a => a.AttributeClass.Is(KnownType.System_Diagnostics_Contracts_PureAttribute)) is { } pureAttribute
                 ? pureAttribute
                 : null;
+
+        private static bool ReturnsVoid(IMethodSymbol methodSymbol) =>
+            methodSymbol.ReturnsVoid && !methodSymbol.Parameters.Any(p => p.RefKind != RefKind.None);
```

Comment:
As far as I can tell, the RefKind comparison should be done also for methods returning tasks.

Can you please add a test case with:
```
[Pure]
Task TaskDoesNotReturn(int input, out int ret)
{
    ret = input;
    return Task.FromResult(input);
}
```

Reply:
You're right. I added both the tests as the implementation for covering it.